### PR TITLE
Add TableState webutil for paginated+filtered+sorted arrays

### PR DESCRIPTION
**Pull request by blueridger**
# Notes
This is the product of my side quest to learn React. It is a state manager for tables built in ReactNodes. It encapsulates Pagination, Sort, and Filtering.

# Design choices
I chose to use a function/type architecture instead of class architecture because it seemed to better align the with React "use..." architecture, and because it results in much more readable syntax.

In opposition to a discussion DL and I was having on a predecessor PR (https://github.com/sourcecred/sourcecred/pull/2520), I chose to sort the original unfiltered table. This is to enable filtering without sorting, making filtering faster, which is important because sorting is a low-frequency mouse-click operation whereas filtering wants to be a high-frequency key-press operation.

# Automated Testing
`yarn test` however this code is **not** tested by unit tests. I find this acceptable because this code is in the front end and manual UI testing is sufficiently indicative of code health. There is also no precedent that I see of unit testing React code.

# Manual Testing
See test plan in https://github.com/sourcecred/sourcecred/pull/2616

If you want to try filtering for yourself, add in the test filter inputs using this code:
```
<input type="text" onChange={(event) => {
        const value = event.target.value;
        ts.createOrUpdateFilterFn("eventName", (ledgerEvent) => ledgerEvent.action.type.replace("_", " ").toLowerCase().includes(value.toLowerCase()))}}/>
<input type="text" onChange={(event) => {
        const value = event.target.value;
        ts.createOrUpdateFilterFn("eventName2", (ledgerEvent) => ledgerEvent.action.type.replace("_", " ").toLowerCase().includes(value.toLowerCase()))
}}/>
```

### CODE REVIEW

File: yarn.lock
Comment (blueridger) [inline yarn.lock:7264]: Not sure why this got added. halp.

File: src/webutil/tableState.js
Comment (hammadj) [inline src/webutil/tableState.js:99]: Would be good to extract the `20` as a constant `DEFAULT_ROWS_PER_PAGE` before the function definition so its not so much a "magic number".

File: src/webutil/tableState.js
Comment (hammadj) [inline src/webutil/tableState.js:78]: Maybe a name like `defaultOptions` or `defaultState` would be more descriptive

File: src/webutil/tableState.js
Comment (hammadj) [inline src/webutil/tableState.js:170]: No need for useMemo here since `.length` isn't computationally intensive, and since its a plain number it will be strict equal between renders as well (e.g. you might need useMemo if you are returning an object that you want to be referentially equal between renders as a performance optimization to prevent re-renders of pure components, but its not needed here and would actually slow it down).

File: src/webutil/tableState.js
Comment (hammadj) [inline src/webutil/tableState.js:159]: Would wrap this in a `useCallback` to prevent creating a new function on every render, which might help prevent re-renders (which is valuable for large list type components)

**Review by hammadj — COMMENTED** [review]
useTableState looks amazing, deserves to be its own open source package haha, super impressive work considering you're learning React! You're lucky you're learning it with Hooks, its a much better dev experience than the old class components where you had to "unlearn" the old way to get used to the new way with Hooks.

Made a bunch of suggestions / nits, overall looking good though.

File: src/webutil/tableState.js
Comment (hammadj) [inline src/webutil/tableState.js:102]: This Map is outside of React's state management, so its likely to cause bugs / issues with re-rendering components when the filter functions change. Its best to keep everything within Reacts state management. Additionally, Map isn't a good fit because when you call `.set()`, the data inside changes but the Map itself remains referentially equal to its previous state, which means React doesn't detect any changes and therefore doesn't re-render the UI.

A better way to do this would be to also put this in useState:
```ts
type FilterFnState = { [key: string]: null | ((T) => boolean) };

// ...
const [filterFns, setFilterFns] = useState<FilterFnState>({});

const createOrUpdateFilterFn = (
     key: string,
     newFilterFn: $Values<FilterFnState>
   ) => {
     setPageIndex(0);
     // if you pass in a function instead of a value, you can get the current filter state as a param and return the new state
     setFilterFns(oldFilters => ({
       ...oldFilters,
       [key]: newFilterFn,
     }));
     // no need to do .slice since .filter returns a new array instead of mutating the input array
     _filter(originalArray);
   };
```

File: src/webutil/tableState.js
Comment (hammadj) [inline src/webutil/tableState.js:116]: Would be better to use .reduce here. Also shouldn't mutate the params, better to initialize a new variable:
```suggestion
    let filteredArray = _array;
    _filterFns.forEach((fn) => {
      if (fn) filteredArray = filteredArray.filter(fn);
    });
```

File: yarn.lock
Comment (hammadj) [reply yarn.lock:7264]: I got that too, seems like someone didn't commit an updated yarn.lock file from some dependency update. Should be fine

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:87]: The idea is that `initialSort` can be optional, but if it's included, all of the subfields must be included.

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:12]: I did find that in my research. The reason I did it this way was because the TableSortLabel can use the lowercase strings as input, which makes this line cleaner:
```
direction={
                    ts.sortName === DATE_SORT[0] ? ts.sortOrder : SortOrders.ASC
                  }
```

I didn't check if it would accept uppercase though. I'll play with it to see.

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:92]: accepted

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:89]: accepted

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:90]: accepted

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:99]: accepted

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:78]: yeah... I'm thinking `initialOptions`

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:170]: I don't fully understand the nuance here because I still don't really understand React under the hood, but I'll take this suggestion and seek more understanding later.

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:159]: I'll look into it.

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:102]: Again, don't really get the nuance, but I'll play with this.

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:116]: accepted

File: yarn.lock
Comment (topocount) [reply yarn.lock:7264]: @blueridger This is from an reverted merge. Rebasing should eliminate it

File: src/ui/components/LedgerViewer.js
Comment (BrianBelhumeur) [inline src/ui/components/LedgerViewer.js:64]: nit: I advocate for named vs positional attributes. E.g.:
```suggestion
  const DATE_SORT = {
      name: "Date", 
      fn: (e) => e.ledgerTimestamp,
  };
```
then you can access them below like:
```
    sortName: DATE_SORT.name,
    sortOrder: SortOrders.DESC,
    sortFn: DATE_SORT.fn,
```

The reason being is it takes less cognitive load to remember positions (though presumably typing takes care of accidentally using the wrong one?) and it's more a descriptive when reading back the code.

**Review by BrianBelhumeur — APPROVED** [review]
@hammadj covered a lot of great points, so I only have a nit to add. Great work @blueridger! You're catching on to React very quickly. 👍

**Review by topocount — COMMENTED** [review]
@blueridger This is awesome! I recommend adding some more to the docstring about use cases and maybe a simple example sketch.

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:12]: Yeah as I guessed, the TableSortLabel wants it lowercase. So we can make the constant names lowercase to add this sugar or we can leave it. I'll probably leave it.

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:99]: actually, I think I will just make this a required field instead, since there is no real case where you can get away with not providing it.

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js]: accepted

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js]: accepted

File: src/ui/components/LedgerViewer.js
Comment (blueridger) [reply src/ui/components/LedgerViewer.js]: accepted

Comment (blueridger) [reply src/ui/components/LedgerViewer.js]: accepted

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js]: accepted

File: src/ui/components/LedgerViewer.js
Comment (hammadj) [inline src/ui/components/LedgerViewer.js:72]: These are both using useState, but the state is never being changed here (no second `setPaginationOptions` or `setDateSort` options assigned in the return value). If the state is never updated, you can just define the value outside of the LedgerViewer component as a constant.

File: src/ui/components/LedgerViewer.js
Comment (hammadj) [inline src/ui/components/LedgerViewer.js:113]:
```suggestion
                  {DATE_SORT.name}
```

File: src/webutil/tableState.js
Comment (hammadj) [inline src/webutil/tableState.js:155]: Usually types are defined outside of function bodies, should move it out for consistency sake and readability.

File: src/webutil/tableState.js
Comment (hammadj) [inline src/webutil/tableState.js:173]: +1 on filtering the sorted array instead of sorting the filtered array, this means that people will be able to change filters without needing to recalculate the sort (opposite isn't true however, updating sort function will trigger filter to run again, but filtering should be cheaper than sorting so not a biggie at all)

File: src/webutil/tableState.js
Comment (hammadj) [inline src/webutil/tableState.js:184]: missing second param here (surprised it didnt get caught by linter, should def fix that)
```suggestion
  }, [pageIndex, rowsPerPage]);
```

File: src/webutil/tableState.js
Comment (hammadj) [inline src/webutil/tableState.js:191]: Also missing array of dependencies in this useCallback. Note that there's a way to not have to include them in here by passing a function into setSortOrder instead, the param will be the current value:

```suggestion
        setSortOrder(currentSortOrder =>  
          currentSortOrder === SortOrders.DESC ? SortOrders.ASC : SortOrders.DESC
        );
```

File: src/webutil/tableState.js
Comment (hammadj) [inline src/webutil/tableState.js:211]: Although this one doesn't have any dependencies on values outside the callback, you still need to pass an empty array as the second param to useCallback in order to indicate that

File: src/webutil/tableState.js
Comment (hammadj) [inline src/webutil/tableState.js:207]: If you specify this inline to `setFilterFns` you shouldn't have to add those type annotations (would be inferred)

File: src/ui/components/LedgerViewer.js
Comment (blueridger) [reply src/ui/components/LedgerViewer.js:72]: aha, accepted

File: src/ui/components/LedgerViewer.js
Comment (blueridger) [reply src/ui/components/LedgerViewer.js:113]: accepted

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:155]: accepted

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:184]: accepted

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:191]: accepted

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:211]: sure

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:207]: Flow throws a fit about not knowing which case to pick due to ambiguity yadda yadda. I did find an improvement that allows me to specify the function inline with only 1 type annotation, though.

**Review by hammadj — APPROVED** [review]
Great work :)

File: src/ui/components/LedgerViewer.js
Comment (decentralion) [inline src/ui/components/LedgerViewer.js:65]: nit: i suggest `deepFreeze`ing so that the defaults can't get modified.

File: src/ui/components/LedgerViewer.js
Comment (decentralion) [inline src/ui/components/LedgerViewer.js:63]: nit: i suggest `deepFreeze`ing so that the defaults can't get modified.

File: src/webutil/tableState.js
Comment (decentralion) [reply src/webutil/tableState.js:11]: For another take on how we could set up options like this, take a look at the Direction const in core/graph:

https://github.com/sourcecred/sourcecred/blob/651db6de017cc74dfc85f16960144d0849fab105/src/core/graph.js#L151-L160

It sounds like in this case we can't take that approach due to the lowercase keys being expected by TableSortLabel. However, I do recommend using deepFreeze -- allowing something to modify these values at runtime would never be a good thing.

File: src/webutil/tableState.js
Comment (decentralion) [inline src/webutil/tableState.js:17]: Nit; wrap to 80 char

File: src/webutil/tableState.js
Comment (decentralion) [inline src/webutil/tableState.js:122]: What do you mean by memoizing the array?

File: src/webutil/tableState.js
Comment (decentralion) [inline src/webutil/tableState.js:139]: It's kind of a code smell to me that the default is dependent on the data. I imagine it could result in issues where e.g. the data is set to have 5 rows, and then the data is updated to have 6 rows, and now the table is only showing the first 5 rows, resulting in user confusion.

That said: do we actually want the default to be "show all data"? Since this is UI code, we can imagine cases where' that's quite bad, e.g. it tries to render 20,000 rows and hangs. Maybe go with a safer default like 25?

File: src/webutil/tableState.js
Comment (decentralion) [inline src/webutil/tableState.js:164]: if we want a perf improvement here, we could just sort with a reversed comparator.
looking at the implementation, we could do this either by:
1. specifying that PluckFns have type `(T) => number` and then doing `sortBy(array, (x) => -sortFn(x))`
or
2. creating a new `sortByReversed` function

either way, we'd save the call to `array.reverse()`. 

probably not needed for now though.

**Review by decentralion — COMMENTED** [review]
Seems like a well designed class. The bit that feels the strangest to me is the way that the sort and filter objects have string names which are load bearing. So we could run into issues: collisions, or cases where the producer renames the string key and not all consumers realize, etc.

Have you considered using [Symbols](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) instead of strings here? Each sort or filter object can have a Symbol. You can still give them names for convenience, but they're guaranteed to be unique, so there will never be collisions. And since they need to be passed by reference, it would be impossible to get a desync where the producer tries to change the key and the consumer doesn't get the memo. Flow would catch it.

Also, it's nice to have some real usage available to review alongside seeing the design. However, I'd prefer for that to be in a separate PR -- they are semantically distinct commits and also merit different review and test plans. The first PR (adding TableState) merits design discussion and detailed code review, and the second (migrating LedgerViewer) would mostly just need regression testing and visual inspection. Having these as two separate review steps helps ensure things don't fall through the cracks.

Finally, it'd be nice to also see an example of the filtering in action, e.g. by adding an event type filter to the ledger viewer. I'd encourage writing it before merging this PR, as another check on the design, but wouldn't block merging this either.

File: src/ui/components/LedgerViewer.js
Comment (blueridger) [reply src/ui/components/LedgerViewer.js]: accepted

File: src/ui/components/LedgerViewer.js
Comment (blueridger) [reply src/ui/components/LedgerViewer.js]: accepted

**Comment by blueridger** [comment]
> Have you considered using [Symbols](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) instead of strings here?

Love it! Didn't know about this.

**Comment by blueridger** [comment]
Pulling out the LedgerViewer changes into a separate PR.

**Comment by blueridger** [comment]
> it'd be nice to also see an example of the filtering in action

Did you see that in the Manual Testing section of this PR, I've already included a functional code snippet of an event type filter, and have included it in the linked video demo? I didn't want to commit it since I think the UX would need hammered out a little more. Does that fulfill your request sufficiently?

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:122]: That it should be referentially equivalent on a re-render, wrapped in a react memo/state. Is there a better way to say this?

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:17]: wrapping all to 80, and adding a vertical ruler to my IDE for the future.

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:139]: This issue is that it is the responsibility of the UI view to create pagination controls, and the TableState cannot coerce the UI view to use pagination. I prefer the unidirectional semantics of TableState offering Sort, Filter, and Pagination and the UI view cherry-picking which it needs.

File: src/webutil/tableState.js
Comment (blueridger) [reply src/webutil/tableState.js:164]: Yeah I had that on my mind, but this way it supports both number and string sorts, which will be really nice from a UX perspective.

File: src/webutil/tableState.js
Comment (blueridger) [inline src/webutil/tableState.js:244]: unfortunately, I was not able to get symbols to work for the filterFns. I had to use `// $FlowFixMe` since symbols are not supported as keys by flow, and even then, it didn't work. I'd like to continue using strings as a good-enough solution, with hopes to maybe figure out why symbols don't work in the future.

**Pull request merged**
